### PR TITLE
Feature/new dependency resolution

### PR DESCRIPTION
**Pull request by scassard**
This is a rework of the uplift dependency resolution algorithm.
The previous version of the algorithm was quite restrictive and assumed that the dependencies of a package did not evolve over time. 

Thus a new algorithm was written based on the [ruby package manager](https://github.com/CocoaPods/Molinillo/blob/master/ARCHITECTURE.md), which is much more permissive with evolving dependencies.

Here is a document explaining how the new algorithm works :
[[EN] Doc uplift dependency resolution.pdf](https://github.com/DragonBox/uplift/files/3568715/EN.Doc.uplift.dependency.resolution.pdf)

### CODE REVIEW

File: Assets/Plugins/Editor/Uplift/UpliftManager.cs [hunk 1]
```diff
@@ -779,11 +777,7 @@ public void UpdatePackage(PackageRepo newer, bool updateDependencies = true, boo
 					{
 						using (TemporaryDirectory td = dependencyPR.Repository.DownloadPackage(dependencyPR.Package))
 						{
-							InstallPackage(dependencyPR.Package,
-										   td,
-										   def,
-										   updateLockfile: updateLockfile
-										  );
+							InstallPackage(dependencyPR.Package, td, null, true);
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/UpliftManager.cs]: are we now ignoring the `updateLockFile` argument?  Any reason why?

File: Assets/Plugins/Editor/Uplift/DependencyResolution/TransitiveDependencySolver.cs [hunk 2]
```diff
@@ -28,7 +28,7 @@
 
 namespace Uplift.DependencyResolution
 {
-	class TransitiveDependencySolver : IDependencySolver
+	class TransitiveDependencySolver //: IDependencySolver
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/TransitiveDependencySolver.cs]: deprecate and/or remove

File: Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyConflicts.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+					{
+						continue;
+					}
+
+					IVersionRequirement tmpVersion = copiedNode.restrictions[key];
+					copiedNode.restrictions[key] = new NoRequirement();
+					try
+					{
+						IVersionRequirement newRequirementForNode = copiedNode.ComputeRequirement();
+						//IVersionRequirement versionRequirementTobeChecked = conflictRaised.requirement.Requirement.RestrictTo(newRequirementForNode);
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyConflicts.cs]: clean

File: Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyGraph.cs [hunk 4]
```diff
@@ -23,12 +23,14 @@
...
 using System.Collections.Generic;
+using System.Text;
+using Uplift.Common;
 using Uplift.Packages;
 using Uplift.Schemas;
 
 namespace Uplift.DependencyResolution
 {
-	class DependencyGraph
+	public class DependencyGraph
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyGraph.cs]: Not necessary to change visibility if we make several of the classes non public.

File: Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs [hunk 5]
```diff
@@ -25,20 +25,33 @@
...
 	public class DependencyNode
 	{
-		protected IVersionRequirement requirement;
 		protected string repository;
 		protected string name;
-		protected List<DependencyNode> dependencies;
 		protected int index;
 		protected int lowlink;
+
+		public Boolean isChildNode = false;
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs]: do these fields need to be public?

File: Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs [hunk 6]
```diff
@@ -136,5 +148,199 @@ public int Lowlink
...
+							DependencyNode parentNode = activated.FindByName(restrictor);
+							DependencyNode childNode = activated.FindByName(dd.Name);
+
+							Debug.Log("Node for " + dd.Name + " already in tree.");
+							if (!parentNode.Dependencies.Contains(childNode))
+							{
+								Debug.Log("Adding " + restrictor + "as parent for " + dd.Name + ".");
+								parentNode.dependencies.Add(childNode);
+							}
+							activated.FindByName(dd.Name).UpdateNodeRequirements(dd, restrictor);
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs]: do we ignore conflicts here voluntarily?

File: Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs [hunk 7]
```diff
@@ -136,5 +148,199 @@ public int Lowlink
 				lowlink = value;
 			}
 		}
+
+		private void UpdateNodeRequirements(DependencyDefinition newRequirements, string restrictor)
+		{
+			Debug.Log("Updating " + this.name + " requirement " + requirement);
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs]: what happens if this is called while the requirements are already conflicting on that node? Is it possible?

File: Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs [hunk 8]
```diff
@@ -136,5 +148,199 @@ public int Lowlink
...
+					Debug.Log("Depends on " + dd.Name);
+					if (dd != null)
+					{
+						if (!activated.Contains(dd.Name))
+						{
+							Debug.Log("Node added for " + dd.Name);
+							DependencyNode childNode = new DependencyNode(dd);
+							Debug.Log(restrictor + " adds first restriction on new dep node " + dd.Name);
+							childNode.restrictions[restrictor] = dd.Requirement;
+							activated.AddDependency(this, childNode); //activated.AddNode(childNode);
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs]: clean

File: Assets/Plugins/Editor/Uplift/DependencyResolution/PossibilitySet.cs [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+	{
+		public string name;
+		public List<PackageRepo> packages = new List<PackageRepo>();
+
+		public PackageRepo GetMostRecentPackage()
+		{
+			//Could also use VersionParser.GreaterThan(string a, string b)
+
+			Version currentVersion = null;
+			Version mostRecentVersionSofar = null;
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/PossibilitySet.cs]: nitpick: Sofar -> SoFar

File: Assets/Plugins/Editor/Uplift/DependencyResolution/PossibilitySet.cs [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+		public static List<PossibilitySet> GetMatchingPossibilities(List<PossibilitySet> possibilities, string requirementName, IVersionRequirement currentRequirement)
+		{
+			List<PossibilitySet> matchingPossibilities = new List<PossibilitySet>();
+			foreach (PossibilitySet possibilitySet in possibilities)
+			{
+				if (possibilitySet.name != requirementName)
+				{
+					continue;
+				}
+				else
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/PossibilitySet.cs]: unecessary else indentation

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+				else
+				{
+					node = dg.FindByName(requested.Name);
+				}
+
+				if (node.restrictions.ContainsKey("initial"))
+				{
+					throw new IncompatibleRequirementException("initial requirements cannot have twice the same requirement");
+				}
+				else
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs]: unecessary else indentation

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+					node.restrictions["initial"] = requested.Requirement;
+				}
+			}
+
+			//Create dependency state for original dependencies
+			LinkedList<DependencyDefinition> currentDependencies = originalDependencies;
+			List<Conflict> conflicts = new List<Conflict>();
+			DependencyState initialState = new DependencyState(currentDependencies,
+																dg,
+																new List<PossibilitySet>(), //possibilities
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs]: lots

of 

empty

space

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Rewinder.cs [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+			List<State> listOfPossibleRewinds = new List<State>();
+
+			if (stack.Count >= 1)
+			{
+				List<DependencyNode> relatedNodes = currentState.activated.FindNodesRelatedToGivenDependency(dependencyName);
+				foreach (DependencyNode node in relatedNodes)
+				{
+					matchingPossibilityStates = matchingPossibilityStates.Concat(listOfStates)
+																		.Where(state => state.GetType() == typeof(PossibilityState))
+																		.Where(state => ((PossibilityState)state).currentRequirement.Name == node.Name)
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/Rewinder.cs]: empty lines

File: Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs [hunk 14]
```diff
@@ -136,5 +148,199 @@ public int Lowlink
...
+			restrictions[restrictor] = newRequirements.Requirement;
+
+			try
+			{
+				this.requirement = requirement.RestrictTo(restrictions[restrictor]);
+			}
+			catch (IncompatibleRequirementException e)
+			{
+				Debug.Log("Incompatible requirement on node");
+				conflictingRequirementOnNode = newRequirements;
```
Comment (lacostej) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyNode.cs]: Here's one thing I haven't understood well. We store those conflicts on the node. We are potentially finding them on several elements on the dependency graph, but seem to only check them on one node. Am I missing something?

File: Assets/Plugins/Editor/Uplift/Testing/DependencyResolution/TestDependencyResolution.cs [hunk 15]
```diff
@@ -0,0 +1,932 @@
...
+using NUnit.Framework;
+using UnityEngine.TestTools;
+using UnityEngine;
+using System.Text;
+using System.Linq;
+
+namespace Uplift.DependencyResolution
+{
+	[TestFixture]
+	public class TestDependencyResolution
```
Comment (niezbop) [inline Assets/Plugins/Editor/Uplift/Testing/DependencyResolution/TestDependencyResolution.cs]: Weird indentation throughout the whole file in `Upset[] expected = {...}` declarations

File: Assets/Plugins/Editor/Uplift/DependencyResolution/NoTransitiveDependenciesSolver.cs [hunk 16]
```diff
@@ -26,7 +26,7 @@
 
 namespace Uplift.DependencyResolution
 {
-	class NoTransitiveDependenciesSolver : IDependencySolver
+	class NoTransitiveDependenciesSolver //: IDependencySolver
```
Comment (niezbop) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/NoTransitiveDependenciesSolver.cs]: Bring back interface support

File: Assets/Plugins/Editor/Uplift/DependencyResolution/PackageListStub.cs [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+				correspondingRepo.Package = pkg;
+				correspondingRepo.Repository = new GithubRepository();
+
+				listPackageRepo.Add(correspondingRepo);
+			}
+			return listPackageRepo;
+		}
+
+
+		override public string ToString()
```
Comment (niezbop) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/PackageListStub.cs]: Cool

File: Assets/Plugins/Editor/Uplift/DependencyResolution/IDependencySolver.cs [hunk 18]
```diff
@@ -22,12 +22,16 @@
...
+using System.Collections.Generic;
+using Uplift.Common;
 using Uplift.Schemas;
 
 namespace Uplift.DependencyResolution
 {
 	interface IDependencySolver
 	{
-		DependencyDefinition[] SolveDependencies(DependencyDefinition[] dependencies);
+		//DependencyDefinition[] SolveDependencies(DependencyDefinition[] dependencies);
```
[the remark this answers is not in the document]
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/IDependencySolver.cs]: They both serve testing purposes.

In my mind we have different options there:
- Create a new interface, deprecate this one and keep the other solvers with the deprecated interface
- Completly overwrite this interface (it's not as if anything depended on the API) and upgrade the other solvers
- Completly overwrite this interface and drop the other ones

I would go for either of the last two. We don't care about deprecation that's going to remain in the code for ever. We want a clean transition

File: Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyGraph.cs [hunk 19]
```diff
@@ -41,6 +43,21 @@ public DependencyGraph(List<DependencyNode> nodes)
 			this.nodeList = nodes;
 		}
 
+		public DependencyGraph(PackageRepo[] packages)
+		{
+			nodeList = new List<DependencyNode>();
+			if (packages != null && packages.Length > 0)
```
[the remark this answers is not in the document]
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/DependencyGraph.cs]: Probably yeah

File: Assets/Plugins/Editor/Uplift/DependencyResolution/PackageListStub.cs (hunk 17 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/PackageListStub.cs]: Although we could try to remain consistent in our method definitions. See `DependencyGraph` that declares it `public`

File: Assets/Plugins/Editor/Uplift/DependencyResolution/PackageListStub.cs [hunk 20]
```diff
@@ -0,0 +1,87 @@
+using System;
```
Comment (niezbop) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/PackageListStub.cs]: License!

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs [hunk 21]
```diff
@@ -0,0 +1,255 @@
...
+		DependencyGraph baseGraph = new DependencyGraph();
+		LinkedList<DependencyDefinition> originalDependencies;
+		Stack<State> stateStack = new Stack<State>();
+
+		public Resolver(DependencyGraph baseGraph, PackageList packageList)
+		{
+
+			this.baseGraph = baseGraph;
+			this.packageList = packageList;
+			if (packageList.GetType() == typeof(PackageListStub))
```
Comment (niezbop) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs]: I don't like this at.

We don't want to have the Resolver know whether the PackageList is a stub or not, we want the classes to offer the same interface and behave differently

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs (hunk 11 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs]: Not sure what you're referring to. C# conventions are not Java's 😄 

https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/if-else

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+					{
+						// Conflicts were found, rewind in a previous state to find solution
+						Conflict conflict = currentState.conflicts.ToArray()[0];
+						Rewinder rewinder = new Rewinder(stateStack);
+						stateStack = rewinder.UnwindForConflict(conflict, packageList);
+						currentState.conflicts.Remove(conflict);
+					}
+					else
+					{
+						// State Resolution is successfull
```
Comment (niezbop) [inline Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs]: Nitpick: successful

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs (hunk 12 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs]: This is github related. These are still one lines, but the tabulation bring them to another line. Look at the `+` at the beginning of the line. They're normal.

Although that may be a tabulation issue then

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs (hunk 11 above)
Comment (lacostej) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs]: For @scassard I meant that you can write
```
 if (x)
 {
    throw ...
 }
 List<Con...
```
and avoid one indentation level. I tend to prefer this (for exceptions or `break/continue` statements) as it makes better use of the editor space.

File: Assets/Plugins/Editor/Uplift/DependencyResolution/IDependencySolver.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (lacostej) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/IDependencySolver.cs]: Agreed.

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs (hunk 12 above)
Comment (lacostej) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/Resolver.cs]: ok now I get why: it's the tabulation and there's code on the right... I hate tabs.

File: Assets/Plugins/Editor/Uplift/DependencyResolution/Rewinder.cs (hunk 13 above)
Comment (niezbop) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/Rewinder.cs]: Tabs as well

Comment (lacostej) [reply Assets/Plugins/Editor/Uplift/DependencyResolution/Rewinder.cs]: ##$)/Y#)" de tabs!!!

**Pull request merged**